src/providers/structure.ts

```diff
@@ -30,31 +31,71 @@ export class SectionNodeProvider implements vscode.TreeDataProvider<Section> {
 
     }
 
-    private refresh(): Section[] {
-        if (this.extension.manager.rootFile) {
-            this.ds = this.buildModel(new Set<string>(), this.extension.manager.rootFile)
+    private async refresh(): Promise<Section[]> {
+        const document = vscode.window.activeTextEditor?.document
+        if (document?.languageId === 'bibtex') {
+            await this.buildBibTeXModel()
+            return this.ds
+        }
+        else if (this.extension.manager.rootFile) {
+            this.ds = this.buildLaTeXModel(new Set<string>(), this.extension.manager.rootFile)
             return this.ds
         } else {
             return []
         }
     }
 
     update() {
-        this.refresh()
-        this._onDidChangeTreeData.fire(undefined)
+        this.refresh().finally(() => {this._onDidChangeTreeData.fire(undefined)})
+    }
+
+    async buildBibTeXModel() {
+        const document = vscode.window.activeTextEditor?.document
+        if (!document) {
+            return
+        }
+        const ast = await this.extension.pegParser.parseBibtex(document.getText())
```

Comment:
Don't we actually need to change https://github.com/James-Yu/LaTeX-Workshop/blob/74a5ae3552872fdcdb0d853477216f55d0bf4bbc/src/components/parser/syntax.ts#L38-L40 into

```ts
async parseBibtex(s: string, options?: bibtexParser.ParserOptions): Promise<bibtexParser.BibtexAst | undefined> {
    return (await this.proxy).parseBibtex(s, options).timeout(30000).catch(() => undefined)
}
```

and check `ast` as

```ts
if (!ast) {
    this.extension.logger.addLogMessage(`Cannot parse a BibTeX file: ${document.filename}`)
    this.ds = []
    return
}

Reply:
Let me have a check. In fact I did not read into `syntax.ts`.